File: sdk/go/tests/secp256k1_test.go
/**
 * Copyright 2017 Intel Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ------------------------------------------------------------------------------
 */

package tests

import (
	. "sawtooth_sdk/signing"
	"testing"
)

var (
	data   = []byte{0x01, 0x02, 0x03}
	PEMSTR = `-----BEGIN EC PRIVATE KEY-----
MHQCAQEEIISGREvlByLRnbhovpK8wSd5hnymtY8hdQCOvZQ473CpoAcGBSuBBAAK
oUQDQgAEWC6TyM1jpYu3f/GGIuktDk4nM1qyOf9PEPHkRkN8zK2HxxNwDi+yN3hR
8Ag+VeTwbRRZOlBdFBsgPxz3/864hw==
-----END EC PRIVATE KEY-----
`
	PEMSTRPRIV = "8486444be50722d19db868be92bcc12779867ca6b58f2175008ebd9438ef70a9"
	ENCPEM     = `-----BEGIN EC PRIVATE KEY-----
Proc-Type: 4,ENCRYPTED
DEK-Info: AES-128-CBC,23CDF282F2217A9334A2413D78DAE04C

PQy89wdLsayP/FG68wgmL1EdlI3S5pN8ibCFrnp5OAtVNrYUD/TH9DMYVmRCNUB4
e+vXoQzd1IysjoFpV21zajSAxCmcbU4CGCDEea3GPwirOSE0ZjPHPp15IkRuGFYm
L/8e9mXvEQPAmBC0NMiltnk4/26iN7hB1QxSQQwy/Zc=
-----END EC PRIVATE KEY-----
`
	ENCPEMPRIV = "2cc32bc33935a5dbad8118abc63dfb627bb91a98d5e6310f5d60f5d65f6adb2f"
	PEMPUBSTR  = "03582e93c8cd63a58bb77ff18622e92d0e4e27335ab239ff4f10f1e446437cccad"
	ENCPEMPUB  = "0257510b4718fd79b21dee3173ffb48ab9a668a35a377be7b7dc432243a940c510"
	PUBSTR     = "035e1de3048a62f9f478440a22fd7655b80f0aac997be963b119ac54b3bfdea3b7"
	SIGSTR     = "0062bc154dca72472e66062c4539c8befb2680d79d59b3cc539dd182ff36072b199adc1118db5fc1884d50cdec9d31a2356af03175439ccb841c7b0e3ae83297"
)

func TestSigning(t *testing.T) {
	context := NewSecp256k1Context()
	priv_1 := context.NewRandomPrivateKey()
	pub_1 := context.GetPublicKey(priv_1)

	sig_1 := context.Sign(data, priv_1)

	if !context.Verify(sig_1, data, pub_1) {
		t.Error(
			"Context fails t to verify signature",
			priv_1, pub_1, sig_1,
		)
	}

	priv_2 := context.NewRandomPrivateKey()

	sig_2 := context.Sign(data, priv_2)

	if context.Verify(sig_2, data, pub_1) {
		t.Error(
			"Context verifies wrong signature",
			priv_2, pub_1, sig_2,
		)
	}

	// Verify that everything returns the right algorithm name
	assertSecp256k1(context.GetAlgorithmName(), t)
	assertSecp256k1(priv_1.GetAlgorithmName(), t)
	assertSecp256k1(pub_1.GetAlgorithmName(), t)
	assertSecp256k1(priv_2.GetAlgorithmName(), t)
}

func assertSecp256k1(name string, t *testing.T) {
	if name != "secp256k1" {
		t.Error("Wrong name", name)
	}
}

func TestPemLoader(t *testing.T) {
	// Load the keys
	priv, err := PemToSecp256k1PrivateKey(PEMSTR, "")
	if err != nil {
		t.Error("Failed to load unencrypted PEM key")
	}

	epriv, err := PemToSecp256k1PrivateKey(ENCPEM, "password")
	if err != nil {
		t.Error("Failed to load encrypted PEM key")
	}

	// Test that they match expected
	if priv.AsHex() != PEMSTRPRIV {
		t.Error("Failed to parse unencrypted PEM key")
	}

	if epriv.AsHex() != ENCPEMPRIV {
		t.Error("Failed to parse encrypted PEM key")
	}

	// Test that the correct public keys are generated
	context := NewSecp256k1Context()

	pub := context.GetPublicKey(priv).AsHex()
	epub := context.GetPublicKey(epriv).AsHex()

	if pub != PEMPUBSTR {
		t.Error("Failed to generate correct public key from unencrypted PEM key")
	}

	if epub != ENCPEMPUB {
		t.Error("Failed to generate correct public key from encrypted PEM key")
	}
}

func TestOtherSigning(t *testing.T) {
	context := CreateContext("secp256k1")
	priv := context.NewRandomPrivateKey()

	factory := NewCryptoFactory(context)
	signer := factory.NewSigner(priv)

	pub := signer.GetPublicKey()
	sig := signer.Sign(data)

	if !context.Verify(sig, data, pub) {
		t.Error(
			"Context fails t to verify signature",
			priv, pub, sig,
		)
	}
}

File: sdk/javascript/spec/signing/secp256k1_test.js
/**
 * Copyright 2017 Intel Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the 'License');
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an 'AS IS' BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * ------------------------------------------------------------------------------
 */

'use strict'

const assert = require('assert')

const { createContext, CryptoFactory, ParseError } = require('../../signing')
const {
  Secp256k1PrivateKey,
  Secp256k1PublicKey
} = require('../../signing/secp256k1')

const KEY1_PRIV_HEX =
    '2f1e7b7a130d7ba9da0068b3bb0ba1d79e7e77110302c9f746c3c2a63fe40088'
const KEY1_PUB_HEX =
    '026a2c795a9776f75464aa3bda3534c3154a6e91b357b1181d3f515110f84b67c5'

const KEY2_PRIV_HEX =
    '51b845c2cdde22fe646148f0b51eaf5feec8c82ee921d5e0cbe7619f3bb9c62d'
const KEY2_PUB_HEX =
    '039c20a66b4ec7995391dbec1d8bb0e2c6e6fd63cd259ed5b877cb4ea98858cf6d'

const MSG1 = 'test'
const MSG1_KEY1_SIG =
  ('5195115d9be2547b720ee74c23dd841842875db6eae1f5da8605b050a49e' +
   '702b4aa83be72ab7e3cb20f17c657011b49f4c8632be2745ba4de79e6aa0' +
   '5da57b35')

const MSG2 = 'test2'
const MSG2_KEY2_SIG =
  ('d589c7b1fa5f8a4c5a389de80ae9582c2f7f2a5e21bab5450b670214e5b1' +
   'c1235e9eb8102fd0ca690a8b42e2c406a682bd57f6daf6e142e5fa4b2c26' +
   'ef40a490')

describe('Secp256k1', () => {
  describe('parsing', () => {
    it('should parse hex keys', () => {
      let privKey = Secp256k1PrivateKey.fromHex(KEY1_PRIV_HEX)
      assert.equal('secp256k1', privKey.getAlgorithmName())
      assert.equal(KEY1_PRIV_HEX, privKey.asHex())

      let pubKey = Secp256k1PublicKey.fromHex(KEY1_PUB_HEX)
      assert.equal('secp256k1', pubKey.getAlgorithmName())
      assert.equal(KEY1_PUB_HEX, pubKey.asHex())
    })

    it('should check invalid digits', () => {
      let privKeyChars = KEY1_PRIV_HEX.replace('1', 'i')
      assert.throws(() => {
        Secp256k1PrivateKey.fromHex(privKeyChars)
      }, ParseError)

      let pubKeyChars = KEY1_PUB_HEX.replace('1', 'i')
      assert.throws(() => {
        Secp256k1PublicKey.fromHex(pubKeyChars)
      }, ParseError)
    })
  })

  describe('private to public', () => {
    it('should return the correct public key', () => {
      let context = createContext('secp256k1')
      assert.equal('secp256k1', context.getAlgorithmName())

      let privKey1 = Secp256k1PrivateKey.fromHex(KEY1_PRIV_HEX)
      assert.equal(KEY1_PRIV_HEX, privKey1.asHex())

      let publicKey1 = context.getPublicKey(privKey1)
      assert.equal(KEY1_PUB_HEX, publicKey1.asHex())

      let privKey2 = Secp256k1PrivateKey.fromHex(KEY2_PRIV_HEX)
      assert.equal(KEY2_PRIV_HEX, privKey2.asHex())

      let publicKey2 = context.getPublicKey(privKey2)
      assert.equal(KEY2_PUB_HEX, publicKey2.asHex())
    })
  })

  describe('signing', () => {
    it('should correcly produce a signature', () => {
      let context = createContext('secp256k1')
      let factory = new CryptoFactory(context)
      assert.equal('secp256k1', factory.getContext().getAlgorithmName())

      let privKey = Secp256k1PrivateKey.fromHex(KEY1_PRIV_HEX)
      let signer = factory.newSigner(privKey)
      let signature = signer.sign(Buffer.from(MSG1))
      assert.equal(MSG1_KEY1_SIG, signature)
    })

    it('should correctly produce multiple signatures', () => {
      let context = createContext('secp256k1')

      let privKey1 = Secp256k1PrivateKey.fromHex(KEY1_PRIV_HEX)
      let privKey2 = Secp256k1PrivateKey.fromHex(KEY2_PRIV_HEX)
      let signature = context.sign(Buffer.from(MSG1), privKey1)
      assert.equal(signature, MSG1_KEY1_SIG)

      signature = context.sign(Buffer.from(MSG2), privKey2)
      assert.equal(signature, MSG2_KEY2_SIG)
    })
  })

  describe('verification', () => {
    it('should succeed on correct public key', () => {
      let context = createContext('secp256k1')

      let pubKey1 = Secp256k1PublicKey.fromHex(KEY1_PUB_HEX)
      assert.equal(pubKey1.getAlgorithmName(), 'secp256k1')
      assert.equal(pubKey1.asHex(), KEY1_PUB_HEX)

      let result = context.verify(MSG1_KEY1_SIG, Buffer.from(MSG1), pubKey1)
      assert.equal(result, true)
    })

    it('should fail on incorrect public key', () => {
      let context = createContext('secp256k1')

      let pubKey1 = Secp256k1PublicKey.fromHex(KEY1_PUB_HEX)
      assert.equal(pubKey1.getAlgorithmName(), 'secp256k1')
      assert.equal(pubKey1.asHex(), KEY1_PUB_HEX)

      // This signature doesn't match for MSG1/KEY1
      let result = context.verify(MSG2_KEY2_SIG, Buffer.from(MSG1), pubKey1)
      assert.equal(result, false)
    })
  })
})
